### PR TITLE
feat(failover): port scrape_bad_nodes to ezpz.failover (PR 1/3)

### PR DESCRIPTION
**Pull request by saforem2**
## Summary

First of three PRs porting the production bad-node failover system from
[`torchtitan/experiments/ezpz/scripts/`](https://github.com/saforem2/torchtitan/tree/ezpz/torchtitan/experiments/ezpz/scripts) into the `ezpz` package proper.

**Scope of this PR**: the log-parsing layer only. The bash orchestration (`failover_lib.sh`) and the `ezpz launch --failover` flag are follow-ups.

## What's here

- `ezpz.failover.scrape_bad_nodes(log_path, *, machine=None)` — public entry point that returns a deduplicated list of bad hostnames in first-seen order. Auto-detects the current machine via `ezpz.get_machine()`; explicit override for tests + cross-cluster postmortems.
- `ezpz.failover.patterns` — per-machine pattern registry. Aurora's PALS shepherd kills and gloo TCP failures bundled by default; other systems register their own via `register_patterns(machine, [BadNodePattern...])`.
- `python -m ezpz.failover <log>` — drop-in CLI replacement for the original `scrape_bad_nodes.py`. Adds an `--explain` mode that breaks down matches per pattern.

## Why this matters

The torchtitan failover system has caught real Aurora hardware failures across many production training runs (jobs **8459818, 8460301, 8463659, 8470102, 8470103, 8479581, 8466848**, others). Two distinct bad-node failure modes are detected:

| Pattern | Source jobs |
|---|---|
| PALS shepherd signal-9 kill (node-local daemon went non-responsive — hardware fault) | 8459818, 8460301, 8463659 |
| gloo TCP peer-closed (IP reverse-resolved via `getent hosts`) | 8470102, 8470103, 8479581 |

The crucial **non-detection** is `rank N died from signal {11,15}` — those are cascading deaths from a primary kill on a *different* node. Including them would falsely tag innocent nodes. Job 8466848 had rank 1304 die from signal 11 as a downstream effect of a `std::bad_alloc` on rank 2413, on a different node — the kill happened on the first node, but the signal-11 message named the second. This is tested as a regression in `test_innocent_rank_signal_11_not_matched`.

## Public API

```python
from ezpz.failover import scrape_bad_nodes

# Auto-detect machine from hostname
hosts = scrape_bad_nodes("/path/to/attempt-1.log")

# Explicit override (tests, cross-cluster postmortems)
hosts = scrape_bad_nodes("/path/to/log", machine="aurora")
```

```bash
# Drop-in CLI replacement for torchtitan's scrape_bad_nodes.py
python -m ezpz.failover /path/to/log
python -m ezpz.failover --machine aurora --explain /path/to/log
```

## Extending to new machines

```python
from ezpz.failover.patterns import BadNodePattern, register_patterns

register_patterns(
    "my-cluster",
    [
        BadNodePattern(
            name="mycluster.fabric_flap",
            extractor=lambda text: [...],
            description="Fabric link reset; node went unreachable.",
        )
    ],
    hostname_normalizer=lambda h: h if h.endswith(".my-cluster.example.com") else None,
)
```

## Sunspot / Polaris / Frontier?

NOT bundled. Sunspot likely shares Aurora's patterns (same PALS shepherd, same XPU fabric) but the full hostname suffix differs and we haven't postmortemed a real Sunspot failure yet to confirm the exact format. Will add a dedicated `sunspot.py` once we have ground truth.

Polaris/Frontier need their own pattern modules (NCCL signatures, SLURM-on-Frontier shapes) — easy follow-up PRs once real failures provide the source material.

## Tests

20 tests in `tests/test_failover_scrape.py`:

- ✅ Both Aurora patterns (shepherd-9 + gloo-closed)
- ✅ The innocent-rank-N exclusion (regression for job 8466848)
- ✅ IP-resolution mocking (no real DNS in unit tests; the IPs in real Aurora logs are HSN-fabric-only)
- ✅ Hostname normalization (`.hostmgmt` → `.hsn` rewrite for downstream PBS-hostfile-match)
- ✅ Dedup + first-seen ordering across many ranks logging the same peer-closed
- ✅ Unresolvable IP / non-Aurora resolved name → silent drop (better than tagging a wrong node from a bogus lookup)
- ✅ Pattern registry: custom registration, re-registration overwrite, `--explain` per-pattern breakdown
- ✅ Edge cases: empty log, clean log, missing log, binary-corrupt log

All 20 pass; **126 total tests pass** across the broader suite (no regressions).

## Test plan

- [x] `pytest tests/test_failover_scrape.py` — 20/20 pass
- [x] `python -m ezpz.failover --explain <sample-log>` — dedup + innocent-rank exclusion work as expected
- [x] Broader regression: `pytest tests/test_history.py tests/test_memory_metrics.py tests/test_launch_watchdog.py tests/test_failover_scrape.py` — 126/126 pass
- [ ] On a real Aurora postmortem log: `python -m ezpz.failover` produces the same output as the existing `scrape_bad_nodes.py` from torchtitan

## Follow-up PRs (planned)

- **PR 2**: port `failover_lib.sh` → `src/ezpz/bin/failover.sh` (the bash orchestration: split nodefile, yeet env, swap in spares, retry loop). Composes with this PR's `scrape_bad_nodes` Python function.
- **PR 3**: wire `--failover` / `--spare-nodes` / `--max-failover-retries` into `ezpz launch` so users get the whole system as a single CLI flag instead of needing to source a bash lib.

Once all three land, the torchtitan copies become thin shims that import from ezpz (or get deleted outright).

### CODE REVIEW

File: src/ezpz/failover/scrape.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        Returns ``[]`` when no patterns matched OR when there is no
+        pattern set registered for the detected machine.
+    """
+    log_path = Path(log_path)
+    if not log_path.is_file():
+        raise FileNotFoundError(f"log file not found: {log_path}")
+
+    if machine is None:
+        machine = _detect_machine()
+    patterns = get_patterns_for_machine(machine)
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/scrape.py]: Fixed in 5820271. Added a `_resolve_machine_key` helper that lowercases both the auto-detect path and the explicit override. Regression test in `test_explicit_machine_arg_also_normalized` confirms `scrape_bad_nodes(log, machine="Aurora")` now hits the registered `"aurora"` patterns (pre-fix it silently returned []).

This was a real production-grade bug — `ezpz.get_machine()` returns title-case strings, and any caller mirroring that style would have silently lost bad-node detection.

File: src/ezpz/failover/scrape.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    if not patterns:
+        # No registered patterns → nothing to do. Returning [] (rather
+        # than raising) lets the caller's blind-rotation fallback fire,
+        # which is the right behavior for unknown failure modes too.
+        return []
+
+    normalize = get_hostname_normalizer(machine)
+    text = log_path.read_text(errors="replace")
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/scrape.py:113]: Addressed via documentation in 5820271 (not the suggested hard cap).

The suggested 512MB ValueError would break existing callers — the torchtitan production failover has been processing real Aurora training logs (some multi-GB) for months without issue. Failover runs on the head node where this is fine.

I did add an explicit comment in the read_text() call documenting the slurp behavior + suggesting `tail -c 1G` for pathological cases:

```python
# NOTE: this reads the entire log into memory. In practice
# postmortem logs we've handled (multi-rank Aurora training,
# hundreds of MB to a couple GB) fit fine on the head node where
# failover runs. If you hit a 10+ GB log, pre-trim it with
# `tail -c 1G` or similar before passing to this function.
```

If a real OOM ever materializes we can switch to mmap/streaming then — the per-pattern `extractor` callable signature already takes `str`, but moving to a per-line stream is a non-breaking change since callers see the same list[str] output.

File: src/ezpz/failover/scrape.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    normalize = get_hostname_normalizer(machine)
+    out: dict[str, list[str]] = {}
+    for pattern in patterns:
+        matches: list[str] = []
+        for raw_host in pattern.extractor(text):
+            host = normalize(raw_host) if normalize else raw_host
+            if host is not None and host not in matches:
+                matches.append(host)
+        out[pattern.name] = matches
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/scrape.py:160]: Fixed in 5820271. Added a parallel `seen: set[str]` for O(1) dedup — exactly the suggested shape.

File: src/ezpz/failover/scrape.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+"failure with no actionable hostname" and apply whatever its
+blind-rotation policy is — silence here does NOT mean the run
+succeeded, only that the patterns couldn't pinpoint a culprit.
+"""
+
+from __future__ import annotations
+
+from pathlib import Path
+from typing import Iterable
+
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/scrape.py:29]: Fixed in 5820271. Removed the dead `Iterable` import.

File: tests/test_failover_scrape.py [hunk 5]
```diff
@@ -0,0 +1,359 @@
...
+        lines = [
+            f"rank {i}: RuntimeError: [..gloo..] Connection closed by peer "
+            f"[10.0.0.42]:{12000 + i}\n"
+            for i in range(30)
+        ]
+        log = _make_log(tmp_path, "".join(lines))
+        hosts = scrape_bad_nodes(log, machine="aurora")
+        assert hosts == ["x4502c1s3b0n0.hsn.cm.aurora.alcf.anl.gov"]
+
+    def test_unresolvable_ip_skipped(self, tmp_path, monkeypatch):
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_failover_scrape.py:177]: Fixed in 5820271. Added `TestReverseResolveIp` with 6 tests covering the real helper:

- success (getent returns a valid line)
- CalledProcessError (IP not in name service)
- TimeoutExpired (slow name service)
- FileNotFoundError (getent missing — common on macOS / alpine containers)
- empty output
- malformed output (IP only, no hostnames)

All exercise `subprocess.check_output` via monkeypatch, no real DNS — keeps the suite deterministic on dev boxes.

File: tests/test_failover_scrape.py [hunk 6]
```diff
@@ -0,0 +1,359 @@
...
+    killed it. Almost always a hardware fault.
+    """
+
+    def test_single_signal_9_extracted(self, tmp_path):
+        log = _make_log(tmp_path, (
+            "Some normal training output\n"
+            "x4502c1s3b0n0.hsn.cm.aurora.alcf.anl.gov: shepherd died from signal 9\n"
+            "More output after\n"
+        ))
+        hosts = scrape_bad_nodes(log, machine="aurora")
```
[the remark this answers is not in the document]
Comment (saforem2) [reply tests/test_failover_scrape.py:57]: Fixed in 5820271. Added `TestAutoDetectMachine` class with three tests covering the `machine=None` path:

- `test_auto_detect_uses_lowercased_ezpz_machine` — monkeypatches `ezpz.get_machine` to return `"Aurora"`, verifies the lookup still finds patterns (caught a related lowercase-normalization bug — see PR #142's other thread on `scrape.py:L88`)
- `test_explicit_machine_arg_also_normalized` — explicit override also lowercased
- `test_auto_detect_handles_get_machine_raising` — graceful empty-list fallback when get_machine raises

File: src/ezpz/failover/patterns/aurora.py [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+# Pattern 2: gloo TCP peer-closed
+#
+# Log line shape (one of many — gloo wraps in RuntimeError):
+#     RuntimeError: [..gloo..] Connection closed by peer [10.0.0.42]:12345
+#
+# Gloo errors typically point at a single peer IP across many
+# "Connection closed" lines (every rank that was talking to the dead
+# node logs its own copy), so deduplication usually collapses to one
+# node. We yield IPs here; the caller resolves them.
+# ---------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/patterns/aurora.py:71]: Fixed in 5820271. Reworded the gloo-pattern comment to match what the code actually does:

```
We reverse-resolve the IP to a hostname here (using the
shared `reverse_resolve_ip` helper) and yield hostnames; the
scraper's downstream normalizer canonicalizes the suffix.
```

File: src/ezpz/failover/scrape.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    """Per-pattern breakdown of what the scraper matched. Not part of
+    the public API — used by the CLI's ``--explain`` mode (when
+    that's added) and by tests that want to verify a specific pattern
+    fired, not just that *some* pattern fired.
+    """
+    log_path = Path(log_path)
+    if machine is None:
+        machine = _detect_machine()
+    patterns = get_patterns_for_machine(machine)
+    text = log_path.read_text(errors="replace")
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/scrape.py:144]: Fixed in 5820271. Same `_resolve_machine_key` helper used by both `scrape_bad_nodes` and `_collect_all_matches_for_debug` so they stay in sync.

File: src/ezpz/failover/patterns/aurora.py [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+      ``x1234c0s0b0n0.hsn.cm.aurora.alcf.anl.gov`` → unchanged
+      ``x1234c0s0b0n0.hostmgmt2042.cm.aurora.alcf.anl.gov`` → HSN form
+      ``some-other-host``                          → None
+    """
+    if _AURORA_HSN_RX.match(host):
+        return host
+    m = _AURORA_NODE_PREFIX_RX.match(host)
+    if m:
+        return f"{m.group(1)}.hsn.cm.aurora.alcf.anl.gov"
+    return None
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/patterns/aurora.py:122]: Fixed in 5820271. Tightened the normalizer to only rewrite the explicit `.hostmgmtNNNN.cm.aurora.alcf.anl.gov` suffix to HSN. Any other suffix on an Aurora-like prefix returns None now — better to drop a match than tag a wrong node based on a speculative rewrite.

Regression test: `test_unknown_suffix_with_aurora_like_prefix_dropped` covers stale /etc/hosts entries and cross-cluster names that happened to share the prefix shape.

File: src/ezpz/failover/patterns/__init__.py [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    for machines that aren't being used."""
+    if machine not in _PATTERNS:
+        # Try to import the per-machine module; it should register itself
+        # at import time. Swallow ImportError for unknown machines so the
+        # caller gets an empty list (and can decide whether to warn).
+        try:
+            importlib.import_module(f"ezpz.failover.patterns.{machine}")
+        except ImportError:
+            pass
+    return _PATTERNS.get(machine, [])
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/patterns/__init__.py:114]: Fixed in 5820271. Narrowed the catch to only ImportError where `.name` matches the per-machine module path itself (the genuine "this module doesn't exist" case). Other ImportErrors (transitive missing deps, syntax errors in known modules) now re-raise.

Two regression tests cover both branches:
- `test_import_error_inside_known_module_surfaces` — simulates a missing transitive dep, asserts re-raise
- `test_import_error_for_unknown_machine_silent` — simulates a genuinely missing per-machine module, asserts silent empty list

File: src/ezpz/failover/patterns/__init__.py [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+    machine: str,
+    patterns: list[BadNodePattern],
+    hostname_normalizer: "Callable[[str], str | None] | None" = None,
+) -> None:
+    """Register a set of patterns + an optional hostname normalizer for
+    a machine. Idempotent — calling twice with the same machine replaces
+    the previous registration (lets tests inject mock patterns)."""
+    _PATTERNS[machine] = list(patterns)
+    if hostname_normalizer is not None:
+        _HOSTNAME_NORMALIZERS[machine] = hostname_normalizer
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/patterns/__init__.py:83]: Fixed in 5820271. `register_patterns` now explicitly clears the normalizer slot when called with `hostname_normalizer=None`. Regression test: `test_re_register_without_normalizer_clears_old_normalizer`. Docstring updated to spell this out.

Codex flagged the same thing as P1 in its issue-level review — addressed both in one place.

File: src/ezpz/failover/patterns/__init__.py [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    """A single failure-mode signature for one machine.
+
+    Args:
+        name: Human-readable identifier (``"aurora.shepherd_signal_9"``).
+            Used in log messages so postmortem reviewers know which
+            pattern fired.
+        extractor: Callable that takes the full log text and yields
+            hostnames or IPs. Yielding IPs is fine — the registry's
+            ``resolve_to_hostname`` step handles reverse-resolution.
+        description: One-line plain-English summary for help text +
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/patterns/__init__.py:51]: Fixed in 5820271. Rewrote `BadNodePattern.extractor` docstring to be accurate:

```
extractor: Callable that takes the full log text and yields
    hostnames (and only hostnames). If the underlying log
    shape gives you IPs (e.g. the gloo "Connection closed by
    peer [IP]:port" line), the extractor is responsible for
    doing the IP→hostname conversion itself — the registry
    does NOT do a resolve step. Use `reverse_resolve_ip`
    from this module to handle the lookup the same way the
    Aurora patterns do.
```

File: src/ezpz/failover/patterns/__init__.py [hunk 13]
```diff
@@ -0,0 +1,147 @@
...
+import subprocess
+from dataclasses import dataclass
+from typing import Callable, Iterable
+
+# Pattern types --------------------------------------------------------------
+
+# A pattern's ``match`` callable receives the FULL log text (not a stream of
+# lines) and yields zero or more bad-host strings (possibly raw IPs that need
+# resolution). Returning an iterable rather than a list keeps memory bounded
+# on very large logs — the caller deduplicates downstream.
```
[the remark this answers is not in the document]
Comment (saforem2) [reply src/ezpz/failover/patterns/__init__.py:31]: Fixed in 5820271. Reworded the comment to clarify that "iterable" only buys us extractor-side streaming of the *match list*, NOT the log text itself. Also updated the `BadNodePattern.extractor` docstring to be honest about what gets read into memory.

**Comment by saforem2** [comment]
@codex Both issues from your review are addressed in 5820271:

**P1 (stale hostname normalizer on re-registration)** — `register_patterns` now explicitly clears the normalizer slot when called with `hostname_normalizer=None`. Regression test `test_re_register_without_normalizer_clears_old_normalizer` covers it.

**P2 (machine override normalization)** — Added a `_resolve_machine_key` helper that lowercases both the auto-detect path and the explicit override before registry lookup. `scrape_bad_nodes(log, machine="Aurora")` now resolves correctly. Two new tests in `TestAutoDetectMachine` cover the auto-detect lowercasing and the explicit-override lowercasing.

Both were real bugs that would have caused silent failover misbehavior on real production logs — thanks for catching them.
